spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,158 @@
+- `Set(Key, Value)`
+
+`Key` and `Value` are assumed to be byte slices; encoding details are left to a later ICS.
+
+Blockchains also need the ability to introspect their own root-of-trust (with `getRootOfTrust`) in order to confirm that the connecting chain has stored the correct one.
+
+#### Subprotocols
+
+Subprotocols are defined as a set of datagram types and a `handleDatagram` function which must be implemented by the state machine of the implementing blockchain. Datagrams must be relayed between chains by an external process. This process is assumed to behave in an arbitrary manner — no safety properties are dependent on its behavior, although progress is generally dependent on the existence of at least one correct relayer process. Further discussion is deferred to [ICS 18: Relayer Algorithms](../spec/ics-18-relayer-algorithms).
+
+IBC subprotocols are reasoned about as interactions between two chains `A` and `B` — there is no prior distinction between these two chains and they are assumed to be executing the same, correct IBC protocol. `A` is simply by convention the chain which goes first in the subprotocol and `B` the chain which goes second. Protocol definitions should generally avoid including `A` and `B` in variable names to avoid confusion (as the chains themselves do not know whether they are `A` or `B` in the protocol).
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+![Opening Handshake](opening_handshake.png)
+
+A correct protocol execution flows as follows:
+
+| Initiator | Datagram          | Chain |
+| --------- | ----------------- | ----- |
+| Actor     | `ConnOpenInit`    | A     |
+| Relayer   | `ConnOpenTry`     | B     |
+| Relayer   | `ConnOpenAck`     | A     |
+| Relayer   | `ConnOpenConfirm` | B     |
+
+At the end of an opening handshake between two chains implementing the subprotocol, the following properties hold:
+- Each chain has each other's correct root-of-trust as originally specified by the initiating actor.
+- The chains have agreed to a shared connection version.
+- Each chain has knowledge of and has agreed to its identifier on the other chain.
+
+##### Header Tracking
+
+The header tracking subprotocol serves to update the root of trust for an open connection.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+![Header Tracking](header_tracking.png)
+
+##### Closing Handshake
+
+The closing handshake protocol serves to cleanly close a connection on two chains.
+
+This subprotocol will likely need to be permissioned to an entity who "owns" the connection on the initiating chain, such as a particular end user, smart contract, or governance mechanism.
+
```

Comment:
Similar question as above, can closing be triggered in the same time?
How are header tracking messages handled after the closing? (I see an assert state==open in the handle function but hopefully it doesn't mean a crash :)
In general it maybe it could help if we add a state switch statement on top of handle functions. And mention valid messages in each state....or a state diagram :)